File: prg2.cpp
/* Write a c++ program to read and write student objects with fixed length records and
	the fields delimited by "|". Implement pack(), unpack(), modify() and search() methods
*/

#include <iostream>
#include <fstream>
#include <conio.h>
#include <string>
#include <process.h>

using namespace std;

fstream fp;
int n = 0;
char fname[10];

class student {
private:
	char usn[10], name[10], branch[5];
public:
	void pack() {
		char buffer[28];
		cout<<"Enter the usn, name and branch: ";
		cin>>usn>>name>>branch;
		strcpy(buffer, usn);
		strcat(buffer, "|");
		strcpy(buffer, name);
		strcat(buffer, "|");
		strcpy(buffer, branch);
		strcat(buffer, "|");
		int len = strlen(buffer);
		while(len < 27) {
			strcat(buffer, "#");
			len++;
		}

		buffer[27] = '\0';
		fp<<buffer<<"\n";
	}

	void unpack(char buffer[]) {
		char *t;
		t = strtok(buffer, "|");
		cout<<"usn: "<<t<<endl;
		t = strtok(NULL, "|");
		cout<<"Name: "<<t<<endl;
		t = strtok(NULL, "|");
		cout<<"Branch: "<<t<<endl;
	}

	void display() {
		char buffer [28];
		int i, j;
		if(n == 0) {
			cout<< "No records to display: ";
			return;
		}

		cout<<"\n File contents are: \n";
		fp.open(fname, ios::in);
		for(int i = 1; i <= n; i++) {
			fp>> buffer;
			unpack(buffer);
			cout<<endl;
		}

		fp.close();
	}

	void search() {
		char buffer[28], temp[28];
		char key[15];
		char *usn;
		int i, j, k, choice;
		cout<<"Enter the USN to search: ";
		cin>>key;
		fp.open(fname, ios::in|ios::out);
		for(int i = 1; i <= n; ++i) {
			fp>>buffer;
			strcpy(temp, buffer);
			usn = strtok(buffer, "|");
			if(strcmp(key, usn) == 0) {
				cout<<"\n Records found: ";
				unpack(buffer);
				cout<<endl;
				cout<<"\n Do you wish to modify? \n";
				cout<<"\n Enter your choice \n1. yes\n2. No\n";
				cin>>choice;
				if(choice == 1) {
					fp.seekp(-27,ios::cur);
					pack();
				}

				fp.close();
				return;
			}
		}
		cout<<"Record not found.";
		fp.close();
		getch();
	}
};

int main() {
	int i, ch;
	student s1;

	cout<<"Enter the filename: ";
	cin>>fname;
	fp.open(fname, ios::out|ios::trunc);
	fp.close();
	for(;;) {
		cout<<"\n1. Insert \n2. Display\n3. Search and Modify \n4. Exit\n";
		cout<<"Enter your choice: ";
		cin>>ch;
		switch(ch) {
			case 1: n++;
							fp.open(fname, ios::app);
							s1.pack();
							fp.close();
							break;
			case 2: s1.display();
							break;
			
			case 3: s1.search();
							break;
			case 4: exit(0);
			default:
					cout<<"invalid choice!\n";
		}
	}
}

File: prg2/prg2.py
import sys


def checkMaxLen(val: str, max: int):
    if len(val) > max:
        print("exceeded maximum limit!")
        sys.exit()

    return val


INSERT = "INSERT"
EDIT = "EDIT"


class Student:
    # private data members:
    __USNMAX = 10
    __NAMMAX = 10
    __BRNMAX = 5

    __moveto = 0

    def __init__(self, filename):
        self.filename = filename
        self.recordsNum = 0
        self.initializeFile()

    def initializeFile(self):
        f = open(self.filename, "w")
        f.close()

    def pack(self, action):
        usn = checkMaxLen(input("enter usn: "), self.__USNMAX)
        name = checkMaxLen(input("enter name: "), self.__NAMMAX)
        branch = checkMaxLen(input("enter branch: "), self.__BRNMAX)

        buffer = usn + "|" + name + "|" + branch + "|"
        if len(buffer) < 28:
            buffer = buffer + "#" * (28 - len(buffer))

        mode = "a" if action == INSERT else "r+"
        with open(self.filename, mode) as f:
            if action == EDIT:
                f.seek(self.__moveto)
            f.write(f"{buffer}\n")
            if action == INSERT:
                self.recordsNum = self.recordsNum + 1

    def unpack(self, record):
        [usn, name, branch, *_] = record.split("|")
        print(f"USN: {usn}\nName: {name}\nBranch: {branch}\n")

    def display(self):
        if self.recordsNum == 0:
            print("No records to display!")
            return

        print("\nFile contents are: ")

        with open(self.filename, "r") as f:
            records = [record.strip() for record in f.readlines()]
            for record in records:
                self.unpack(record)

    def search(self):
        key = checkMaxLen(input("enter usn to search: "), self.__USNMAX)

        with open(self.filename, "r+") as f:
            for i in range(self.recordsNum):
                record = f.readline()
                [usn, *_] = record.split("|")

                if usn == key:
                    print("Record Found!")
                    self.unpack(record)

                    ch = int(input("Do you wish to modify? enter 1 to do so: "))
                    if ch == 1:
                        self.__moveto = (
                            f.tell() - 30
                        )  # substract 30(including newline) character from curr pos to move to the line beginning
                        self.pack(EDIT)

                    return

            print("Record not found!")


if __name__ == "__main__":

    filename = input("Enter the file name: ")
    stu = Student(filename)

    while True:
        ch = int(
            input(
                "1. Insert\n2. Display\n3. Search and Modify\n4. Exit\nEnter your choice: "
            )
        )
        if ch == 1:
            stu.pack(INSERT)
        elif ch == 2:
            stu.display()
        elif ch == 3:
            stu.search()
        elif ch == 4:
            break
        else:
            print("Invalid choice!")
